Feature/tcd biomasse wrsi

caldwellst opened the pull request:
Here's the draft code for loading, processing, and comparing Biomasse and WRSI.

Tinkaa commented on src/indicators/drought/biomasse.py:
```diff
@@ -0,0 +1,335 @@
+data and then recalculates cumulative biomasse and anomaly
+using that. Since DMP measurements are productivity values, the
+calculation of total biomasse in a year is taken by averaging
+the DMP over the entire period and multiplying by 365.25.
```
Aah this took me a while to get. But is it correct that the dekadal DMP is the mean daily production during that dekad? I first interpreted it as the total production during that dekad. But if the first is correct then the calculation of the total biomasse makes sense to me. If so, maybe you could clarify that here?

caldwellst replied:
Yeah, you ar ecorrect, have clarified in the documentation!

Tinkaa commented on src/indicators/drought/biomasse.py:
```diff
@@ -0,0 +1,335 @@
+    )
+
+    # calculate anomaly for mean separate from
+    # observed to ensure unique dekadal values
```
this comment is vague to me. Season index is basically a shifted dekad number such that 1 corresponds to the start of the season right?

Tinkaa commented on src/indicators/drought/biomasse.py:
```diff
@@ -0,0 +1,335 @@
+    df_merged.sort_values(
+        by=admin_cols + ["year", "dekad"], inplace=True, ignore_index=True
+    )
+    df_merged["season"] = df_merged["year"] + (
```
I am a bit confused by the definition of season, I would expect it to be one lower.. Cause now e.g. the 2022 season starts in the `start_dekad` of 2021, so on 1 April 2021. It is not super bad but would at least add a comment explaining that. And be careful to later on not mix up season definition

caldwellst replied:
Yeah, I will subtract from 1 so it's clear, but just added it here for the grouping, it wasn't actually included in the output file. However, have decided probably good to add to the output as well, so will make it even clearer by having it be "year1-year2".

Tinkaa commented on src/indicators/drought/biomasse.py:
```diff
@@ -0,0 +1,335 @@
+        )
+        .reset_index()
+        .drop(labels="IDBIOHYDRO", axis=1)
+        .groupby(admin_cols + ["time"])
```
you might want to check here first that there are no duplicated `[admin_cols]+["time"]` values? Since we saw this occur once? And now you are summing the values in case there is a duplicate, which I am doubtful whether that is what we should do. (especially since you are taking the mean after the groupby on `df_mean`

Tinkaa replied:
Aha so this is the same issue with the polygons being split into two. Just make sure that the method after aggregation (I think it should be sum?) is the same with the dmp and mean data

caldwellst replied:
Yes, making sure they are all aligned! We take the mean in the first instance because we want to get a single measure of productivity for each unique area, and summing would be an overstimate. However, now we have a measure of productivity for each area so we want to sum it across our time period.

Tinkaa replied:
@caldwellst I am not sure if I agree with this. I get the mean and the sum in general but when there is duplicate areas, I don't think it is correct. Maybe I am wrong but want to make sure we are not doing smth wrong. 

For example in Liberia we have the admin2 area with pcode `LR0402`. When I use the processed file (`biomasse_general_ADM2_dekad_10.csv`) and check the anomalies for that pcode, I see that they are always 180 or higher. If I understand the anomaly correctly the average anomaly should be 100 and thus it is not correct that it is always higher than 100. I would think the reason is that for the mean you take the mean over the two noncontigunous polygons while for the dekad-yearly DMP you take the sum. What is your take on this?

caldwellst replied:
Aha yes I see now. I think moving to a weighted average for the DMP values themselves makes sense then. I will run this and test if it's comparable. Thanks for the catch here!

Tinkaa commented on src/indicators/drought/biomasse.py:
```diff
@@ -0,0 +1,335 @@
+    ].apply(lambda x: x.cumsum() * 365.25 / 36)
+    df_merged["biomasse_anomaly"] = (
+        100 * df_merged["biomasse"] / df_merged["biomasse_mean"]
+    )
```
All the calculations look good to me :)

Tinkaa commented on src/indicators/drought/biomasse.py:
```diff
@@ -0,0 +1,335 @@
+_processed_filename = "biomasse_{iso3}_{admin}_dekad_{start_dekad}.csv"
+
+
+def _get_processed_path(admin, start_dekad, iso3: Union[None, str] = None):
```
Commented somewhere else, but could also use `glb` instead of `general`

Tinkaa commented on src/indicators/drought/biomasse.py:
```diff
@@ -0,0 +1,335 @@
+    _check_admin(admin)
+    df = load_biomasse_data(admin=admin, start_dekad=start_dekad)
+    admin_col = admin.replace("ADM", "admin") + "Pcod"
+    df_subset = df[df[admin_col].isin(admin_pcodes)]
```
maybe add a comment here on descripttion that the exact boundaries of the admin shapes that Biomasse uses are not known? So solely concat based on pcode?

Tinkaa replied:
In practice shouldn't give a large difference but think good to note somewhere as a caveat

caldwellst replied:
In this case, we are concating based on Pcode because it's the easiest way to join, but they are the same geographic areas as area of interest. There are some areas outside our area of interest that don't match at the ADMIN2 level but the ADMIN2's completely align for our area of interest in this case.

Tinkaa commented on src/indicators/drought/biomasse.py:
```diff
@@ -0,0 +1,335 @@
+    admin_col = admin.replace("ADM", "admin") + "Pcod"
+    df_subset = df[df[admin_col].isin(admin_pcodes)]
+    # just keep to year 2000 for unique value per dekad
+    # when calculating biomasse mean aggregated
```
maybe add in comment that you select just one year (=2000) since the mean in all years is the same

caldwellst replied:
Good point!

Tinkaa commented on src/indicators/drought/biomasse.py:
```diff
@@ -0,0 +1,335 @@
+    # sorting data again year then dekad
+    df_merged.sort_values(by=["year", "dekad"], inplace=True)
+
+    if iso3 is not None:
```
why would you only save it if you have an `iso3`? Maybe in this case do not allow the `iso3` to be `None`?

caldwellst replied:
In this case @Tinkaa, my thought process was you might want to use the function without saving automatically to a specific country file. Maybe if you were exploring for instance cross-country analysis? Then you would need to manually save because we don't have an immediately logical filename and folder in our structure for it to fit into. What do you think? Previously I actually had it as a filepath argument, but realized in general we are wanting to save to an `iso3` folder.

Tinkaa replied:
Hmm yeah I have been thinking about this as well.. I figured that if the case occurs of cross-country analysis we might as well come up with a fake iso3. So basically that the iso3 arg just serves as an ID. Though that might get a bit ambiguous, especially if we want to check for validity of the iso3 in the toolbox. 

It is just that if we add the None option for iso3 we would basically have to check for that with every dataset that covers more than a country before saving. Which might be a bit ugly. 

However, for now it is fine to leave it in. The thing I would change though then is to make the iso3 a mandatory argument instead of having the default as None. But then keep the option for now to be None instead of a str. Would that make sense? 

@turnerm please chime in if you got any thoughts on this topic beyond this specific case.

caldwellst replied:
I like that idea, let's see what @turnerm thinks.

Tinkaa commented on src/indicators/drought/biomasse.py:
```diff
@@ -0,0 +1,335 @@
+
+def load_biomasse_mean(mask: bool = True):
+    """Used to mask WRSI"""
+    bm_fp = os.path.join(_raw_path, "BiomassValueMean.tif")
```
where do you get this file from?

Tinkaa commented on src/indicators/drought/biomasse.py:
```diff
@@ -0,0 +1,335 @@
+    bm_fp = os.path.join(_raw_path, "BiomassValueMean.tif")
+    da = rioxarray.open_rasterio(bm_fp)
+    if mask:
+        da = da.where(da.values <= 0, drop=True)
```
I get an error when running `load_biomasse_mean()`. Seems to be caused by the `drop=True`. Assuming you didn't finalize this function yet but fyi

caldwellst replied:
Yes good catch!

Tinkaa reviewed (commented):
I did a thorough review of `biomasse.py` and while I left some comments all computations look good! Let me know if it is ready to not be draft anymore/other files need a draft review.

caldwellst commented on src/indicators/drought/biomasse.py:
```diff
@@ -0,0 +1,335 @@
+
+from src.utils_general.utils import download_ftp
+
+_base_url = (
```
Will put the global constants in caps, very good point! I put the `_` in front to signal they shouldn't be exported (I think in python modules if the underscore is in front, then doing `import module`, you can never access them using `module._constant`, you can only access explicitly using `from module import _constant`. Do you know if that's right @Tinkaa @turnerm?

Tinkaa replied:
btw I learnt from monica that yes. It should have an underscore and capitals

caldwellst commented on src/indicators/drought/biomasse.py:
```diff
@@ -0,0 +1,335 @@
+_raw_filename = "WA_DMP_{admin}_ef_v0.csv"
+
+_raw_path = os.path.join(
+    os.getenv("AA_DATA_DIR"), "public", "raw", "general", "biomasse"
```
It does make sense @Tinkaa, I just totally missed that this existed cause haven't used it previously.

caldwellst commented on src/indicators/drought/biomasse.py:
```diff
@@ -0,0 +1,335 @@
+    df_dmp = df.filter(regex="(^admin|^DMP_[0-9]+|^ID)")
+    admin_cols = [col for col in df_mean.columns if col.startswith("admin")]
+
+    # groupby to average out for the few cases where admin areas
```
Happens there and in some other areas in Côte d'Ivoire, have mentioned. It's not a duplicate administrative area, rather that the administrative area is not contiguous so is split into two separate polygons in GIS systems, hence why the output contains multiple rows for them. Isn't a problem at all for us but figured should solve in this manner just in case we eventually use data for those areas.

caldwellst replied:
Have amended the comment and code to fit your comments, thanks!

Tinkaa replied:
Aaah now I get what you mean, thanks!

caldwellst commented:
I've addressed all your comments @Tinkaa except the ones unresolved, thanks a lot for them! I also went through the NDVI and WRSI files to make those comments also fit there. As well, NDVI processing, downloading, and analysis has been completed and is also part of the PR. Was easier to do all in one PR since most of the work was comparing Biomasse. The rest is ready for review, I just need to go through on Friday and update the documentation of the functions.

Tinkaa commented on src/indicators/drought/wrsi.py:
```diff
@@ -0,0 +1,309 @@
+
+
+def filter_wrsi(da, dekad):
+    dt = _dekad_to_date(2001, dekad)
```
aha so here you select the data from one dekad across all years? Would another option to have an extra coordinate named `dekad` (which is dependent on time) such that you can just filter on dekad. 

If you never use the `time` directly it could also be an option to instead have a `year` and `dekad` dimension but can imagine it is also easy to have them combined into one.

caldwellst replied:
Yeah, I have added as a TODO. The reason it was troublesome before, cause I tried, was beacuse then you have multiple dimensions and then I was using rioxarray and that was problematic. Then i switched it all up, a mess. Something to investigate soon, this great!

Tinkaa commented on src/indicators/drought/wrsi.py:
```diff
@@ -0,0 +1,309 @@
+    return year, dekad
+
+
+def _dekad_to_date(year: int, dekad: int):
```
even though private function I would add short docstring if the computed date represents the beginning of the dekad (would assume so) or e.g. the end

caldwellst replied:
Yes, didn't document this well yet, but going through now!

Tinkaa commented on src/indicators/drought/wrsi.py:
```diff
@@ -0,0 +1,309 @@
+
+
+def _wrsi_percent_below_thresh(da, threshold):
+    da = xr.where(da <= threshold, 1, 0).mean(dim=["x", "y"])
```
nice!

Tinkaa reviewed (commented):
Alrighty reviewed `wrsi.py` Seems like many comments but most are small, great work once again! 

Will continue with reviewing the other files

Tinkaa commented on src/indicators/drought/ndvi.py:
```diff
@@ -0,0 +1,192 @@
+    processed_path.unlink(missing_ok=True)
+    df.to_csv(processed_path)
+
+    return df
```
Could you also add a load function?

Tinkaa commented on src/indicators/drought/ndvi.py:
```diff
@@ -0,0 +1,192 @@
+        date_obj = date.fromisoformat(date_obj)
+    year = date_obj.year
+    dekad = (
+        (date_obj.timetuple().tm_mday // 10) + ((date_obj.month - 1) * 3) + 1
```
Maybe I am misinterpreting but why would you use `date_obj.timetuple().tm_mday` instead of `date_obj.day` ? Also applies to the same function in `wrsi.py`

Tinkaa commented on analyses/tcd/notebooks/forecasts/tcd_ndvi.md:
````diff
@@ -0,0 +1,85 @@
+```python
````
Could you explain a bit what you are doing?

Tinkaa commented on src/indicators/drought/ndvi.py:
```diff
@@ -0,0 +1,192 @@
+    )
+
+
+def process_ndvi(iso3, geometries, thresholds):
```
this does still take quite long to do. Which isn't bad but would add a comment about it. Plus maybe a progress bar could be an idea?

caldwellst replied:
Yeah agreed.

Tinkaa commented on analyses/tcd/notebooks/forecasts/tcd_wrsi.md:
````diff
@@ -0,0 +1,212 @@
+
+```python
+d = wrsi_percent_below(da_range_anom_mask, [70, 80,90,100,110,120])
+d.to_csv(os.path.join(_save_processed_path, "tcd_wrsi_anomaly_rangeland_thresholds.csv"))
````
comment out saving paths

Tinkaa commented on analyses/tcd/notebooks/forecasts/tcd_ndvi.md:
```diff
@@ -0,0 +1,85 @@
+from rasterio.enums import Resampling
+import hvplot.xarray
+import altair as alt
+
```
There is also lots of libraries that are imported but not used.

Tinkaa commented on analyses/tcd/notebooks/forecasts/tcd_ndvi.md:
````diff
@@ -0,0 +1,85 @@
+```python
+gdf_adm1=gpd.read_file(adm1_bound_path)
+gdf_reg=gdf_adm1[gdf_adm1.area_of_interest == True]
+
````
I would also add the function call of `download_ndvi()` and comment it out. But just so the whole process is documented.

Tinkaa commented on analyses/tcd/notebooks/forecasts/tcd_ndvi.md:
````diff
@@ -0,0 +1,85 @@
+```
+
+```python
+df1 = rioxarray.open_rasterio(os.path.join(config.DATA_DIR, "public", "raw", "glb", "ndvi", "wa0333pct.tif"))
````
What are you doing here and below? Almost seems you are testing the methodology that is now implemented in `process_ndvi`?

caldwellst replied:
Ha yes, was testing so have removed now!

Tinkaa commented on analyses/tcd/notebooks/forecasts/tcd_wrsi.md:
```diff
@@ -0,0 +1,212 @@
+
+from pathlib import Path
+import sys
+import os
```
remove unused libraries

Tinkaa commented on analyses/tcd/notebooks/forecasts/tcd_wrsi.md:
```diff
@@ -0,0 +1,212 @@
+This notebook explores the WRSI trigger related to Chad. Given that the WRSI is not a forecast based method (although the extended version does include historical precipitation averages), I have been unable to find any evaluations of its accuracy or skill. We will thus dive straight into understanding how the historical patterns look in Chad and what a reasonable threshold and trigger might look  ...
+
```
could you add that further exploration is done in the R file?

Tinkaa commented on analyses/tcd/notebooks/forecasts/tcd_biomasse.md:
````diff
@@ -0,0 +1,40 @@
+import src.indicators.drought.biomasse as bm
+```
+
+This is just the code for downloading and procesing the Biomasse data, then aggregating to a specific set of admin codes (our region of interest in Chad.
````
would add that further exploration is done in the R file

Tinkaa commented on src/indicators/drought/biomasse.py:
```diff
@@ -0,0 +1,362 @@
+
+def load_dmp(admin: AdminArgument = "ADM2", redownload: bool = False):
+    _check_admin(admin)
+    raw_path = _get_raw_path(admin)
```
hmm don't you want to check here for file existence instead of redownloading every time you load the dataset?

caldwellst replied:
Was all a bit of a mixup switching to `unlink`, woops!

Tinkaa commented on src/indicators/drought/biomasse.py:
```diff
@@ -0,0 +1,362 @@
+    also pivoting the mean values to allow for
+    calculation of biomasse anomaly.
+    """
+    df = load_dmp(admin, redownload)
```
I would assume the dataset is already downloaded when calling process. So check for file existence and if doesn't exist raise an error.

Tinkaa commented on src/indicators/drought/biomasse.py:
```diff
@@ -0,0 +1,362 @@
+    return None
+
+
+def load_dmp(admin: AdminArgument = "ADM2", redownload: bool = False):
```
the redownload arg is never used..

Tinkaa reviewed (commented):
Alrighty reviewed everything except for the R scripts. 

And all the computations look good to me, woohoo. 

Only comments on code style and usability. Some things I commented earlier on the wrsi script I also saw in the ndvi script, so can do a second pass once we discussed those comments.

caldwellst commented on src/indicators/drought/ndvi.py:
```diff
@@ -0,0 +1,192 @@
+"""Code to download and process NDVI data
+
+NDVI data is downloadable and available through the USGS
```
Hmm not sure I get why? As in explaining that it's smoothed/slightly processed already?

Tinkaa replied:
Yeah that is what I meant but very minor. Just from a toolbox perspective that people might expect it is the raw NDVI data, but we would possibly explain a bit more in that case anyway

caldwellst commented on src/indicators/drought/wrsi.py:
```diff
@@ -0,0 +1,309 @@
+    else:
+        start_year, start_dekad = _date_to_dekad(start_date)
+
+    if isinstance(end_date, list):
```
The `end_date` structure is slightly amended just because if it isn't a `list`, then both methods need to call `_date_to_dekad()`, so wanted to not repeat myslf by calling it twice. But maybe still clearer to just match and do some slight repetition? What do you think?

Tinkaa replied:
Aah I see, missed that. Then yes this is fine. Could maybe with the start_date to first `if list` and then `elif is None` to mimic the structure a bit more, but that is very minor.

caldwellst commented on src/indicators/drought/wrsi.py:
```diff
@@ -0,0 +1,309 @@
+        end_year, end_dekad = _date_to_dekad(end_date)
+
+    i_year = start_year
+    i_dekad = min(start_dekad, 33)
```
So we could restrict the user passed values ,but they might just want to pass in like `2000-01-01` to `2010-12-31` or something to download all dates. And default `end_date` is just `date.today()` which I think is easiest, so wanted to just limited those so we don't get errors. Have explained more in documentation now.

caldwellst commented on src/indicators/drought/wrsi.py:
```diff
@@ -0,0 +1,309 @@
+    raw_dir = _get_raw_dir(region)
+    raw_dir.mkdir(parents=True, exist_ok=True)
+
+    if not redownload:
```
Okay clear! I saw this once so kept using it but had no other logic than that hahahaah. I will see how the `toolbox` is using it.

caldwellst replied:
`clobber` sounds cool as shit hahaha. I will go with that. Also makes it quicker to transfer to `pa-aa-toolbox`.

caldwellst commented on src/indicators/drought/wrsi.py:
```diff
@@ -0,0 +1,309 @@
+    raw_dir = _get_raw_dir(region)
+    raw_dir.mkdir(parents=True, exist_ok=True)
+
+    if not redownload:
```
Ugh, you're not wrong, so much smoother. I just got into my head and did this because it was slightly more complex with ARC2 where we storing data within a single file, rather than the multiple files here. Refactoring!!!

caldwellst replied:
So much shorter and cleaner. NICE.

caldwellst commented on src/indicators/drought/wrsi.py:
```diff
@@ -0,0 +1,309 @@
+    try:
+        url = _get_url_path(year=year, dekad=dekad, region=region)
+        resp = urlopen(url)
+    except HTTPError:
```
So, I don't actually want to raise an error here, I should actually have as `logger.message`. In the previous instance of the data (pre-CHIRPS), there were certain missing years/dekads that weren't downloaded because they didn't exist, but isn't an actual error, just important to catch and log for the user if they might want to query why some data is missing. I put `logger.error` because I was avoiding changing my setting to show messages by default in console lol...But should probably be message so have changed! Hopefully isn't an issue anymore anyway with the CHIRPS data.

caldwellst commented on src/indicators/drought/wrsi.py:
```diff
@@ -0,0 +1,309 @@
+    for type in ["do", "eo", "er"]:
+        file_name = _BASE_FILENAME.format(year=year, dekad=dekad, type=type)
+        save_path = Path(raw_dir, file_name)
+        save_path.unlink(missing_ok=True)
```
No need to overwrite. I think I kept it in because I saw old scripts talking about issues with overwriting, and just mindlessly kept that!

caldwellst commented on src/indicators/drought/wrsi.py:
```diff
@@ -0,0 +1,309 @@
+def load_wrsi(region, type):
+    processed_path = _get_processed_path(region, type)
+    a = xr.open_dataset(processed_path).to_array("wrsi")
+    a = a.squeeze(drop=True)
```
Yeah, I think it's nicer to explicitly save as dataset, because `DataArray.to_netcdf()` converts to a dataset on the backend, and wasn't sure if there was a better format to save in instead of `NetCDF` (prefer that to saving just a million band `.tif` for instance). But yes, will open with `rioxarray` and see how it's different! These comments are gold, thanks so much.

caldwellst commented on src/indicators/drought/wrsi.py:
```diff
@@ -0,0 +1,309 @@
+
+_BASE_FILENAME = "w{year:04}{dekad}{type}.tif"
+
+
```
Done!!

caldwellst commented on src/indicators/drought/ndvi.py:
```diff
@@ -0,0 +1,192 @@
+    i_year = start_year
+    i_dekad = start_dekad
+
+    if not redownload:
```
So much smoother now.

caldwellst commented on src/indicators/drought/ndvi.py:
```diff
@@ -0,0 +1,192 @@
+
+def _get_paths(year, dekad):
+    base_file_name = _BASE_FILENAME.format(year=year % 100, dekad=dekad)
+    url_zip_filename = base_file_name + ".zip"
```
f-string fiend.

caldwellst commented:
Madness. So many good comments, thanks Tinka! Have hopefully addressed almost all, just a couple left with TODO to be done when merging into `pa-aa-toolbox`. Love this continuous learning and thanks again for all the helpful tips.

Tinkaa commented on src/indicators/drought/biomasse.py:
```diff
@@ -0,0 +1,432 @@
+
+
+def _get_processed_path(
+    admin: AdminArgument, start_dekad: int, iso3: Union[None, str] = None
```
No need to change, but for learning: the recommended way is to instead of using Union with None to use `Optional` see [here](https://stackoverflow.com/questions/39429526/how-to-specify-nullable-return-type-with-type-hints)

caldwellst replied:
Brilliant, thank you!!

Tinkaa commented on src/indicators/drought/wrsi.py:
```diff
@@ -0,0 +1,382 @@
+should be considered and we should be careful in learning
+to apply and use both products where applications.
+
+I have also just found the CHIRPS WRSI data on the USGS website.
```
Is this still relevant?

Tinkaa commented on src/indicators/drought/wrsi.py:
```diff
@@ -0,0 +1,382 @@
+            end_date = date.today()
+        end_year, end_dekad = _date_to_dekad(end_date)
+
+    i_year = max(start_year, 2001)
```
I would move this check already up to the `start_year,start_dekad` determination and thus add an extra line with `start_year = max(start_year,2001)`

Tinkaa commented on src/indicators/drought/wrsi.py:
```diff
@@ -0,0 +1,382 @@
+        end_year, end_dekad = _date_to_dekad(end_date)
+
+    i_year = max(start_year, 2001)
+    i_dekad = max(min(start_dekad, 33), 13)
```
same here, `start_dekad=...`

Tinkaa commented on src/indicators/drought/wrsi.py:
```diff
@@ -0,0 +1,382 @@
+        for filename in raw_dir.glob("*.tif")
+    ]
+    while not (
+        i_year > end_year or (i_year == end_year and i_dekad > i_dekad)
```
Sorry I thought I commented this before but apparently I didn't. I would recommend using a for loop with a zip over the years and dekads instead of the `while not`. So smth like `for year,dekad in zip(range(min_year,max_year),range(13,33):`

Only for the start_year you then would have to have one check like `if year==start_year and dekad<start_dekad:continue` I would say that is cleaner and prevents you from doing the plus one and checking if dekad is larger than 33 at the end. Does that make sense to you as well?

Tinkaa replied:
Also applies to the ndvi `while not` code

Tinkaa commented on src/indicators/drought/wrsi.py:
```diff
@@ -0,0 +1,382 @@
+    to around 17, and from the eastern border of Chad
+    to Senegal. This is specifically for rangelands:
+    https://earlywarning.usgs.gov/fews/product/57
+
```
You mention it later but would also add here that data is only available from dekad 13 till 33 (add dates as well)

Tinkaa reviewed (commented):
Super super great work! Added a few more comments to the py files (sorry!) that I only catched now. But think then it is ready to go

Two general comments: 
- would be good to clean-up old files on the GDrive. Saw for example the `general` folder still existing
- When I was playing a little with the Biomasse data, I saw some negative anomaly values. Which were cause by negative Biomasse (mean) values so would be good to double-check that and see what that means.

caldwellst commented:
Thanks for all the comments! Have resolved them all and hope they are good to go now. Regarding some of your above comments.

- I removed `general` from my machine before leaving, and still don't see it here. Not sure why still on drive, but have removed, hopefully that fully cleans it up.

- Good catch on negative values. I thought the missing values were all -9998.8 but there were also -9999.0 which I wasn't catching. All seems good now when checking, and have used averaging across both the `DMP_MEA` and `DMP` values. Now the checks seem to work out.

I have messaged back to the Biomasse team at ACF so we can discuss with them the methods applied above and confirm that we aren't doing anything wrong.

Tinkaa commented:
You are amazing, thanks! Last thing I did was try to run the R scripts and only two minor changes to make them run, so committed them. 

One last thing is that we discussed before that it would be nice to compare our own computed biomasse anomaly to the one produced by the platform. Do you agree that that would still be a good sanity check? If so, I think it makes sense to do that in a separate PR and thus merge this one 🎉

caldwellst commented:
Yes, good idea! Will need to think through how to do that and then try and get that done. Will merge this then do that.
